File: scripts_legacy/plot_tewl_missing_heatmap.py
from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd
from PIL import Image, ImageDraw, ImageFont


ROOT = Path(__file__).resolve().parents[1]
INPUT = ROOT / "processed_data" / "04_tewameter" / "TEWL_processed_raw_window.xlsx"
OUTPUT = ROOT / "processed_data" / "04_tewameter" / "TEWL_final_missing_heatmap.png"


def font(size: int, bold: bool = False) -> ImageFont.FreeTypeFont | ImageFont.ImageFont:
    candidates = [
        "/System/Library/Fonts/Supplemental/Arial Bold.ttf" if bold else "/System/Library/Fonts/Supplemental/Arial.ttf",
        "/System/Library/Fonts/Supplemental/Helvetica Bold.ttf" if bold else "/System/Library/Fonts/Supplemental/Helvetica.ttf",
    ]
    for candidate in candidates:
        try:
            return ImageFont.truetype(candidate, size)
        except OSError:
            pass
    return ImageFont.load_default()


def clean_blank(value):
    if isinstance(value, str) and value.strip().lower() in {"", "na", "n/a"}:
        return np.nan
    return value


def draw_rotated_label(base: Image.Image, xy: tuple[int, int], text: str, text_font, fill=(45, 45, 45)) -> None:
    label = Image.new("RGBA", (130, 32), (255, 255, 255, 0))
    draw = ImageDraw.Draw(label)
    draw.text((0, 0), text, font=text_font, fill=fill)
    rotated = label.rotate(90, expand=True)
    base.alpha_composite(rotated, xy)


def main() -> None:
    df = pd.read_excel(INPUT, sheet_name="RawWindowSummaryClean")
    df["final_clean_tewl"] = df["final_clean_tewl"].map(clean_blank)
    df["real day"] = pd.to_numeric(df["real day"], errors="coerce")
    df["subject"] = df["subject"].astype(str)

    stage_order = (
        df[["real stage", "real day"]]
        .drop_duplicates()
        .sort_values(["real day", "real stage"], kind="stable")
        .reset_index(drop=True)
    )
    stage_keys = list(stage_order.itertuples(index=False, name=None))
    stage_labels = [str(stage) for stage, _day in stage_keys]

    groups = ["Control", "Ex", "ExAG"]
    group_subjects = {
        group: sorted(df.loc[df["group"] == group, "subject"].dropna().unique())
        for group in groups
    }

    cell = 26
    left = 110
    group_gap = 34
    label_h = 18
    bottom_label_h = 165
    legend_h = 70
    title_h = 54
    cols = len(stage_keys)
    rows_total = sum(len(group_subjects[group]) for group in groups) + group_gap // cell * (len(groups) - 1)
    width = left + cols * cell + 40
    height = (
        title_h
        + label_h
        + sum(len(group_subjects[group]) * cell for group in groups)
        + group_gap * (len(groups) - 1)
        + bottom_label_h
        + legend_h
    )

    img = Image.new("RGBA", (width, height), "white")
    draw = ImageDraw.Draw(img)
    title_font = font(19, bold=True)
    label_font = font(10)
    small_font = font(9)
    group_font = font(12, bold=True)

    draw.text((18, 16), "TEWL final_clean_tewl Missingness by Subject, Group, Stage", font=title_font, fill=(20, 20, 20))

    present_color = (232, 243, 255, 255)
    missing_color = (201, 52, 47, 255)
    grid_color = (255, 255, 255, 255)
    text_color = (40, 40, 40)

    total_missing = 0
    total_cells = 0
    y = title_h + label_h
    for group in groups:
        subjects = group_subjects[group]
        matrix = np.ones((len(subjects), cols), dtype=int)
        for i, subject in enumerate(subjects):
            sub = df[(df["group"] == group) & (df["subject"] == subject)]
            for j, (stage, day) in enumerate(stage_keys):
                row = sub[(sub["real stage"] == stage) & (sub["real day"] == day)]
                if not row.empty and row["final_clean_tewl"].notna().any():
                    matrix[i, j] = 0

        missing_count = int(matrix.sum())
        total_missing += missing_count
        total_cells += matrix.size

        draw.text((18, y + max(0, len(subjects) * cell // 2 - 10)), group, font=group_font, fill=text_color)
        draw.text((18, y + max(0, len(subjects) * cell // 2 + 8)), f"{missing_count}/{matrix.size}", font=small_font, fill=(90, 90, 90))
        for i, subject in enumerate(subjects):
            row_y = y + i * cell
            draw.text((72, row_y + 6), subject, font=label_font, fill=text_color)
            for j in range(cols):
                x = left + j * cell
                color = missing_color if matrix[i, j] else present_color
                draw.rectangle((x, row_y, x + cell - 1, row_y + cell - 1), fill=color)
                draw.rectangle((x, row_y, x + cell - 1, row_y + cell - 1), outline=grid_color)
        y += len(subjects) * cell + group_gap

    stage_label_y = y - group_gap + 12
    for j, label in enumerate(stage_labels):
        draw_rotated_label(img, (left + j * cell - 2, stage_label_y), label, small_font, text_color)

    legend_y = stage_label_y + 138
    draw.rectangle((left, legend_y, left + 18, legend_y + 18), fill=present_color, outline=(180, 180, 180))
    draw.text((left + 24, legend_y + 2), "Present", font=label_font, fill=text_color)
    draw.rectangle((left + 110, legend_y, left + 128, legend_y + 18), fill=missing_color, outline=(180, 180, 180))
    draw.text((left + 134, legend_y + 2), "Missing or no record", font=label_font, fill=text_color)
    draw.text(
        (18, height - 20),
        f"Source: {INPUT.name} / RawWindowSummaryClean. Total missing/no-record cells: {total_missing}/{total_cells}.",
        font=small_font,
        fill=(80, 80, 80),
    )

    img.convert("RGB").save(OUTPUT, quality=95)
    print(OUTPUT)


if __name__ == "__main__":
    main()

File: scripts_legacy/plot_tewl_missing_heatmap.R
library(ggplot2)
library(readxl)

args_file <- normalizePath(sub("^--file=", "", grep("^--file=", commandArgs(FALSE), value = TRUE)[1]), mustWork = TRUE)
root <- normalizePath(file.path(dirname(args_file), ".."), mustWork = TRUE)
input <- file.path(root, "processed_data", "04_tewameter", "TEWL_processed_raw_window.xlsx")
output <- file.path(root, "processed_data", "04_tewameter", "TEWL_final_missing_heatmap_R.png")

df <- read_excel(input, sheet = "RawWindowSummaryClean")

to_missing <- function(x) {
  ifelse(is.na(x) | tolower(trimws(as.character(x))) %in% c("", "na", "n/a"), NA, x)
}

df$final_clean_tewl <- to_missing(df$final_clean_tewl)
df$subject <- as.character(df$subject)
df$group <- factor(df$group, levels = c("Control", "Ex", "ExAG"))
df$plot_stage <- as.character(df[[4]]) # Excel column D: stage

stage_order <- data.frame(plot_stage = unique(df$plot_stage), stringsAsFactors = FALSE)
stage_order$stage_id <- as.character(seq_len(nrow(stage_order)))
stage_order$stage_label <- stage_order$plot_stage

subjects <- do.call(
  rbind,
  lapply(levels(df$group), function(g) {
    data.frame(group = g, subject = sort(unique(df$subject[df$group == g])), stringsAsFactors = FALSE)
  })
)
subjects$subject_id <- ave(subjects$subject, subjects$group, FUN = seq_along)

grid <- merge(subjects, stage_order, by = NULL)
observed <- df[, c("group", "subject", "plot_stage", "final_clean_tewl", "Rik_score")]
names(observed) <- c("group", "subject", "plot_stage", "final_clean_tewl", "Rik_score")
observed$present_value <- !is.na(observed$final_clean_tewl)
present_lookup <- aggregate(
  present_value ~ group + subject + plot_stage,
  data = observed,
  FUN = any
)
score_lookup <- aggregate(
  Rik_score ~ group + subject + plot_stage,
  data = observed,
  FUN = function(x) {
    non_missing <- x[!is.na(x)]
    if (length(non_missing) == 0) NA else non_missing[1]
  },
  na.action = na.pass
)
observed <- merge(present_lookup, score_lookup, by = c("group", "subject", "plot_stage"), all = TRUE)
plot_df <- merge(grid, observed, by = c("group", "subject", "plot_stage"), all.x = TRUE)
plot_df$Rik_score <- suppressWarnings(as.numeric(plot_df$Rik_score))
plot_df$status <- ifelse(
  is.na(plot_df$present_value) | !plot_df$present_value,
  "Missing or no record",
  "Present"
)
plot_df$status <- factor(
  plot_df$status,
  levels = c("Missing or no record", "Present")
)

missing_summary <- aggregate(
  missing ~ group,
  data = transform(plot_df, missing = status == "Missing or no record"),
  FUN = sum
)
total_summary <- aggregate(stage_id ~ group, data = plot_df, FUN = length)
names(total_summary)[2] <- "total"
group_labels <- merge(missing_summary, total_summary, by = "group")
group_labels$label <- paste0(group_labels$group, "\n", group_labels$missing, "/", group_labels$total)

plot_df$subject <- factor(plot_df$subject, levels = rev(unique(subjects$subject)))
plot_df$stage_id <- factor(plot_df$stage_id, levels = stage_order$stage_id)

p <- ggplot(plot_df, aes(x = stage_id, y = subject, fill = status)) +
  geom_tile(color = "white", linewidth = 0.35) +
  facet_grid(group ~ ., scales = "free_y", space = "free_y", switch = "y") +
  scale_x_discrete(labels = setNames(stage_order$stage_label, stage_order$stage_id)) +
  scale_fill_manual(
    values = c(
      "Missing or no record" = "#c9342f",
      "Present" = "#e8f3ff"
    ),
    drop = FALSE
  ) +
  labs(
    title = "TEWL final_clean_tewl Missingness by Subject, Group, Stage",
    x = NULL,
    y = NULL,
    fill = NULL,
    caption = paste0("Source: ", basename(input), " / RawWindowSummaryClean")
  ) +
  theme_minimal(base_size = 10) +
  theme(
    panel.grid = element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 8),
    axis.text.y = element_text(size = 8),
    strip.placement = "outside",
    strip.text.y.left = element_text(angle = 0, face = "bold", size = 10),
    legend.position = "bottom",
    plot.title = element_text(face = "bold", size = 14),
    plot.caption = element_text(size = 7, hjust = 0),
    panel.spacing.y = unit(0.6, "lines")
  )

ggsave(output, p, width = 9, height = 10.5, dpi = 300)
cat(output, "\n")
